Mathlib/GroupTheory/GroupAction/DomAct.lean

```diff
@@ -0,0 +1,219 @@
+/-
+Copyright (c) 2023 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+import Mathlib.Algebra.Group.Opposite
+import Mathlib.Algebra.Group.Pi
+import Mathlib.GroupTheory.GroupAction.Defs
+
+/-!
+# Type tags for right action on the domain of a function
+
+By default, `M` acts on `α → β` if it acts on `β`, and the action is given by
+`(c • f) a = c • (f a)`.
+
+In some cases, it is useful to consider another action: if `M` acts on `α` on the left, then it acts
+on `α → β` on the right so that `(c • f) a = f (c • a)`. E.g., this action is used to reformulate
+the Mean Ergodic Theorem in terms of an operator on \(L^2\).
+
+## Main definitions
+
+- `DomMulAct M` (notation: `Mᵈᵐᵃ`): type synonym for `Mᵐᵒᵖ`; if `M` multiplicatively acts on `α`,
+  then `Mᵈᵐᵃ` acts on `α → β` for any type `β`;
+- `DomAddAct M` (notation: `Mᵈᵃᵃ`): the additive version.
+
+We also define actions of `Mᵈᵐᵃ` on:
+
+- `α → β` provided that `M` acts on `α`;
+- `A →* B` provided that `M` acts on `A` by a `MulDistribMulAction`;
+- `A →+ B` provided that `M` acts on `A` by a `DistribMulAction`.
+
+## Implementation details
+
+### Motivation
+
+Right action can be represented in `mathlib` as an action of the opposite group `Mᵐᵒᵖ`. However,
+this "domain shift" action cannot be an instance because this would create a "diamond"
+(a.k.a. ambigous notation): if `M` is a monoid, then how does `Mᵐᵒᵖ` act on `M → M`? On the one
```

Comment:
```suggestion
(a.k.a. ambiguous notation): if `M` is a monoid, then how does `Mᵐᵒᵖ` act on `M → M`? On the one
```

Reply:
I mean, the short answer is: write `DomMulAct.mk c⁻¹ • f`. Otherwise, these two actions have the same orbits etc.